README.md
scandir, a better directory iterator that returns all file info the OS provides
===============================================================================

scandir is a module which provides a generator version of `os.listdir()` that
also exposes the extra file information the operating system returns when you
iterate a directory. scandir also provides a much faster version of
`os.walk()`, because it can use the extra file information exposed by the
scandir() function.


Background
----------

Python's built-in `os.walk()` is significantly slower than it needs to be,
because -- in addition to calling `listdir()` on each directory -- it calls
`stat()` on each file to determine whether the filename is a directory or not.
But both `FindFirstFile` / `FindNextFile` on Windows and `readdir` on Linux/OS
X/BSD already tell you whether the files returned are directories or not, so
no further `stat` system calls are needed. In short, you can reduce the number
of system calls from about 2N to N, where N is the total number of files and
directories in the tree.

**In practice, removing all those extra system calls makes `os.walk()` about
8-9 times as fast on Windows, and about 3-4 times as fast on Linux and Mac OS
X.** So we're not talking about micro-optimizations. [See more benchmarks
below.](#benchmarks)

Somewhat relatedly, many people have also asked for a version of
`os.listdir()` that yields filenames as it iterates instead of returning them
as one big list. This improves memory efficiency for iterating very large
directories.

So as well as a faster `walk()`, scandir adds a new `scandir()` function.
They're pretty easy to use, but [see below](#the-api) for the full API docs.


Why you should care
-------------------

I'd love for these incremental (but significant!) improvements to be added to
the Python standard library. This scandir module was released to help test the
concept and get it in shape for inclusion in the standard `os` module.

There are various third-party "path" and "walk directory" libraries available,
but Python's `os` module isn't going away anytime soon. So we might as well
speed it up and add small improvements where possible.

**So I'd love it if you could help test scandir, report bugs, suggest
improvements, or comment on the API.** And perhaps you'll see these speed-ups
and API additions in Python 3.4 ... :-)


Benchmarks
----------

Below are results showing how many times as fast `scandir.walk()` is than
`os.walk()` on various systems, found by running `benchmark.py` with no
arguments as well as with the `-s` argument (which totals the directory size).

```
System version              Python version    Speed ratio    With -s
--------------------------------------------------------------------
Windows 7 64 bit            2.7 64 bit        8.2            13.8
Windows XP 32 bit           2.7 32 bit

Ubuntu 10.04 32 bit         2.7 32 bit        3.2            1.8

Mac OS X 10.7.5             2.7 64 bit
```

These benchmarks are using the fast C version of scandir_helper, which is
equivalent to os.listdir(), also written in C, except that it provides the
file information along with the name.

Note that the gains are less than the above on smaller directories and greater
on larger directories. This is why `benchmark.py` creates a test directory
tree with a standardized size.


The API
-------

### walk()

The API for `scandir.walk()` is exactly the same as `os.walk()`, so just [read
the Python docs](http://docs.python.org/2/library/os.html#os.walk).

### scandir()

The `scandir()` function is the scandir module's main workhorse. It's defined
as follows:

```python
scandir(path='.') -> iterator of DirEntry objects
```

It yields a DirEntry for each file and directory in `path`. Like os.listdir(),
`.` and `..` are skipped, and the entries are yielded in system-dependent
order. Each DirEntry object has the following attributes and methods:

* `name`: filename, relative to path (like that returned by os.listdir)
* `dirent`: a dirent object or None on systems that don't support it (Windows);
  the dirent object has `d_ino` and `d_type` attributes
* `isdir()`: like os.path.isdir(), but free on most systems (Linux, Windows, OS X)
* `isfile()`: like os.path.isfile(), but free on most systems (Linux, Windows, OS X)
* `islink()`: like os.path.islink(), but free on most systems (Linux, Windows, OS X)
* `lstat()`: like os.lstat(), but free on Windows

Here's a good usage pattern for `scandir`. This is in fact almost exactly how
the faster `os.walk()` implementation uses it:

```python
dirs = []
nondirs = []
for entry in scandir(path):
    if entry.isdir():
        dirs.append(entry)
    else:
        nondirs.append(entry)
```


Further reading
---------------

* [Thread I started on the python-ideas list about speeding up os.walk()](http://mail.python.org/pipermail/python-ideas/2012-November/017770.html)
* [Python Issue 11406, original proposal for scandir(), a generator without the dirent/stat info](http://bugs.python.org/issue11406)
* [Further thread I started on python-dev that refined the scandir() API](http://mail.python.org/pipermail/python-dev/2013-May/126119.html)
* [Question on StackOverflow about why os.walk() is slow and pointers to fix it](http://stackoverflow.com/questions/2485719/very-quickly-getting-total-size-of-folder)
* [Question on StackOverflow asking about iterating over a directory](http://stackoverflow.com/questions/4403598/list-files-in-a-folder-as-a-stream-to-begin-process-immediately)
* [BetterWalk, my previous attempt at this, on which this code is based](https://github.com/benhoyt/betterwalk)

To-do
-----

* Make _scandir.scandir_helper functions return real iterators instead of lists
* Move building of DirEntry objects into C module
* Fix tests on Python 3, especially for [reparse points / Win32 symbolic links](http://mail.python.org/pipermail/python-ideas/2012-November/017794.html)


Flames, comments, bug reports
-----------------------------

Please send flames, comments, and questions about scandir to Ben Hoyt:

http://benhoyt.com/

File bug reports or feature requests at the GitHub project page:

https://github.com/benhoyt/scandir


CHANGES.txt

2012-11-19 version 0.6

* Added fields keyword arg to iterdir_stat() API, per post on python-ideas
  mailing list.
* Added pattern keyword arg for fnmatch-style matching (to make use of
  FindFirst/Next wildcard matching).
* Other fixes and improvements.


2012-11-13 version 0.5

* Initial upload to GitHub


betterwalk.py
"""BetterWalk, a better and faster os.walk() for Python.

BetterWalk is a somewhat better and significantly faster version of Python's
os.walk(), as well as a generator version of os.listdir(). See README.md or
https://github.com/benhoyt/betterwalk for rationale and documentation.

BetterWalk is released under the new BSD 3-clause license. See LICENSE.txt for
the full license text.

"""

import ctypes
import fnmatch
import os
import stat
import sys

__version__ = '0.6'
__all__ = ['iterdir', 'iterdir_stat', 'walk']


# Windows implementation
if sys.platform == 'win32':
    from ctypes import wintypes

    # Various constants from windows.h
    INVALID_HANDLE_VALUE = ctypes.c_void_p(-1).value
    ERROR_FILE_NOT_FOUND = 2
    ERROR_NO_MORE_FILES = 18
    FILE_ATTRIBUTE_READONLY = 1
    FILE_ATTRIBUTE_DIRECTORY = 16
    FILE_ATTRIBUTE_REPARSE_POINT = 1024

    # Numer of seconds between 1601-01-01 and 1970-01-01
    SECONDS_BETWEEN_EPOCHS = 11644473600

    kernel32 = ctypes.windll.kernel32

    # ctypes wrappers for (wide string versions of) FindFirstFile,
    # FindNextFile, and FindClose
    FindFirstFile = kernel32.FindFirstFileW
    FindFirstFile.argtypes = [
        wintypes.LPCWSTR,
        ctypes.POINTER(wintypes.WIN32_FIND_DATAW),
    ]
    FindFirstFile.restype = wintypes.HANDLE

    FindNextFile = kernel32.FindNextFileW
    FindNextFile.argtypes = [
        wintypes.HANDLE,
        ctypes.POINTER(wintypes.WIN32_FIND_DATAW),
    ]
    FindNextFile.restype = wintypes.BOOL

    FindClose = kernel32.FindClose
    FindClose.argtypes = [wintypes.HANDLE]
    FindClose.restype = wintypes.BOOL

    # The conversion functions below are taken more or less straight from
    # CPython's Modules/posixmodule.c

    def attributes_to_mode(attributes):
        """Convert Win32 dwFileAttributes to st_mode."""
        mode = 0
        if attributes & FILE_ATTRIBUTE_DIRECTORY:
            mode |= stat.S_IFDIR | 0o111
        else:
            mode |= stat.S_IFREG
        if attributes & FILE_ATTRIBUTE_READONLY:
            mode |= 0o444
        else:
            mode |= 0o666
        if attributes & FILE_ATTRIBUTE_REPARSE_POINT:
            mode |= stat.S_IFLNK
        return mode

    def filetime_to_time(filetime):
        """Convert Win32 FILETIME to time since Unix epoch in seconds."""
        total = filetime.dwHighDateTime << 32 | filetime.dwLowDateTime
        return total / 10000000.0 - SECONDS_BETWEEN_EPOCHS

    def find_data_to_stat(data):
        """Convert Win32 FIND_DATA struct to stat_result."""
        st_mode = attributes_to_mode(data.dwFileAttributes)
        st_size = data.nFileSizeHigh << 32 | data.nFileSizeLow
        st_atime = filetime_to_time(data.ftLastAccessTime)
        st_mtime = filetime_to_time(data.ftLastWriteTime)
        st_ctime = filetime_to_time(data.ftCreationTime)
        # These are set to zero rather than None, per CPython's posixmodule.c
        st_ino = 0
        st_dev = 0
        st_nlink = 0
        st_uid = 0
        st_gid = 0
        return os.stat_result((st_mode, st_ino, st_dev, st_nlink, st_uid,
                               st_gid, st_size, st_atime, st_mtime, st_ctime))

    def win_error(error, filename):
        exc = WindowsError(error, ctypes.FormatError(error))
        exc.filename = filename
        return exc

    def iterdir_stat(path='.', pattern='*', fields=None):
        """See iterdir_stat.__doc__ below for docstring."""
        # We can ignore "fields" in Windows, as FindFirst/Next gives full stat

        if '[' in pattern or pattern.endswith('?'):
            # Windows FindFirst/Next doesn't support bracket matching, and it
            # doesn't handle ? at the end patterns as per fnmatch; use fnmatch
            wildcard = '*'
        else:
            # Otherwise use built-in FindFirst/Next wildcard matching
            wildcard = pattern
            pattern = None

        # Call FindFirstFile and handle errors
        data = wintypes.WIN32_FIND_DATAW()
        data_p = ctypes.byref(data)
        filename = os.path.join(path, wildcard)
        handle = FindFirstFile(filename, data_p)
        if handle == INVALID_HANDLE_VALUE:
            error = ctypes.GetLastError()
            if error == ERROR_FILE_NOT_FOUND:
                # No files, don't yield anything
                return
            raise win_error(error, path)

        # Call FindNextFile in a loop, stopping when no more files
        try:
            while True:
                # Skip '.' and '..' (current and parent directory), but
                # otherwise yield (filename, stat_result) tuple
                name = data.cFileName
                if name not in ('.', '..'):
                    if pattern is None or fnmatch.fnmatch(name, pattern):
                        st = find_data_to_stat(data)
                        yield (name, st)

                success = FindNextFile(handle, data_p)
                if not success:
                    error = ctypes.GetLastError()
                    if error == ERROR_NO_MORE_FILES:
                        break
                    raise win_error(error, path)
        finally:
            if not FindClose(handle):
                raise win_error(ctypes.GetLastError(), path)

    try:
        import _betterwalk
        iterdir_stat = _betterwalk.listdir
        print('Using fast _betterwalk')
    except ImportError:
        print('CANNOT IMPORT _betterwalk')

# Linux, OS X, and BSD implementation
elif sys.platform.startswith(('linux', 'darwin')) or 'bsd' in sys.platform:
    import ctypes.util

    DIR_p = ctypes.c_void_p

    # Rather annoying how the dirent struct is slightly different on each
    # platform. The only fields we care about are d_name and d_type.
    class dirent(ctypes.Structure):
        if sys.platform.startswith('linux'):
            _fields_ = (
                ('d_ino', ctypes.c_ulong),
                ('d_off', ctypes.c_long),
                ('d_reclen', ctypes.c_ushort),
                ('d_type', ctypes.c_byte),
                ('d_name', ctypes.c_char * 256),
            )
        else:
            _fields_ = (
                ('d_ino', ctypes.c_uint32),  # must be uint32, not ulong
                ('d_reclen', ctypes.c_ushort),
                ('d_type', ctypes.c_byte),
                ('d_namlen', ctypes.c_byte),
                ('d_name', ctypes.c_char * 256),
            )

    DT_UNKNOWN = 0

    dirent_p = ctypes.POINTER(dirent)
    dirent_pp = ctypes.POINTER(dirent_p)

    libc = ctypes.CDLL(ctypes.util.find_library('c'), use_errno=True)
    opendir = libc.opendir
    opendir.argtypes = [ctypes.c_char_p]
    opendir.restype = DIR_p

    readdir_r = libc.readdir_r
    readdir_r.argtypes = [DIR_p, dirent_p, dirent_pp]
    readdir_r.restype = ctypes.c_int

    closedir = libc.closedir
    closedir.argtypes = [DIR_p]
    closedir.restype = ctypes.c_int

    file_system_encoding = sys.getfilesystemencoding()

    def type_to_stat(d_type):
        """Convert dirent.d_type value to stat_result."""
        st_mode = d_type << 12
        return os.stat_result((st_mode,) + (None,) * 9)

    def posix_error(filename):
        errno = ctypes.get_errno()
        exc = OSError(errno, os.strerror(errno))
        exc.filename = filename
        return exc

    def iterdir_stat(path='.', pattern='*', fields=None):
        """See iterdir_stat.__doc__ below for docstring."""
        # If we need more than just st_mode_type (dirent.d_type), we need to
        # call stat() on each file
        need_stat = fields is not None and set(fields) != set(['st_mode_type'])

        dir_p = opendir(path.encode(file_system_encoding))
        if not dir_p:
            raise posix_error(path)
        try:
            entry = dirent()
            result = dirent_p()
            while True:
                if readdir_r(dir_p, entry, result):
                    raise posix_error(path)
                if not result:
                    break
                name = entry.d_name.decode(file_system_encoding)
                if name not in ('.', '..'):
                    if pattern == '*' or fnmatch.fnmatch(name, pattern):
                        if need_stat or entry.d_type == DT_UNKNOWN:
                            st = os.stat(os.path.join(path, name))
                        else:
                            st = type_to_stat(entry.d_type)
                        yield (name, st)
        finally:
            if closedir(dir_p):
                raise posix_error(path)


# Some other system -- have to fall back to using os.listdir() and os.stat()
else:
    def iterdir_stat(path='.', pattern='*', fields=None):
        """See iterdir_stat.__doc__ below for docstring."""
        names = os.listdir(path)
        if pattern != '*':
            names = fnmatch.filter(names, pattern)
        for name in names:
            if fields is not None:
                st = os.stat(os.path.join(path, name))
            else:
                st = os.stat_result((None,) * 10)
            yield (name, st)


# iterdir_stat.__doc__ = """
# Yield tuples of (filename, stat_result) for each filename that matches
# "pattern" in the directory given by "path". Like os.listdir(), '.' and '..'
# are skipped, and the values are yielded in system-dependent order.

# Pattern matching is done as per fnmatch.fnmatch(), but is more efficient if
# the system's directory iteration supports pattern matching (like Windows).

# The "fields" parameter specifies which fields to provide in each stat_result.
# If None, only the fields the operating system can get "for free" are present
# in stat_result. Otherwise "fields" must be an iterable of 'st_*' attribute
# names that the caller wants in each stat_result. The only special attribute
# name is 'st_mode_type', which means the type bits in the st_mode field.

# In practice, all fields are provided for free on Windows; whereas only the
# st_mode_type information is provided for free on Linux, Mac OS X, and BSD.
# """


def iterdir(path='.', pattern='*'):
    """Like iterdir_stat(), but only yield the filenames."""
    for name, st in iterdir_stat(path, pattern=pattern):
        yield name


def walk(top, topdown=True, onerror=None, followlinks=False):
    """Just like os.walk(), but faster, as it uses iterdir_stat internally."""
    # Determine which are files and which are directories
    dirs = []
    dir_stats = []
    nondirs = []
    try:
        for name, st in iterdir_stat(unicode(top)):
            if stat.S_ISDIR(st.st_mode):
                dirs.append(name)
                dir_stats.append(st)
            else:
                nondirs.append(name)
    except OSError as err:
        if onerror is not None:
            onerror(err)
        return

    # Yield before recursion if going top down
    if topdown:
        yield top, dirs, nondirs

    # Recurse into sub-directories, following symbolic links if "followlinks"
    for name, st in zip(dirs, dir_stats):
        new_path = os.path.join(top, name)
        if followlinks or not stat.S_ISLNK(st.st_mode):
            for x in walk(new_path, topdown, onerror, followlinks):
                yield x

    # Yield before recursion if going bottom up
    if not topdown:
        yield top, dirs, nondirs
